### ISSUE TITLE
Child with id 36 is invisible and should not be added to tree.

### ISSUE THREAD

**Issue by lishuhao** [flutter/flutter#13326]
## Exception
```
══╡ EXCEPTION CAUGHT BY SCHEDULER LIBRARY ╞═════════════════════════════════════════════════════════
The following assertion was thrown during a scheduler callback:
Child with id 36 is invisible and should not be added to tree.
'package:flutter/src/semantics/semantics.dart': Failed assertion: line 370 pos 16:
'!child.isInvisible'

Either the assertion indicates an error in the framework itself, or we should provide substantially
more information in this error message to help you determine and fix the underlying cause.
In either case, please report this assertion by filing a bug on GitHub:
https://github.com/flutter/flutter/issues/new

When the exception was thrown, this was the stack:
════════════════════════════════════════════════════════════════════════════════════════════════════
```

## flutter analyze
```
Analyzing /Users/hao/Documents/flutter_proj/shop...
No issues found!
Ran in 5.8s
```

## Flutter Doctor
```
[✓] Flutter (on Mac OS X 10.13.1 17B1003, locale zh-Hans-CN, channel alpha)
    • Flutter at /Users/hao/Documents/flutter
    • Framework revision d957c8f040 (3 days ago), 2017-11-30 13:29:59 -0800
    • Engine revision 77d8acb9be
    • Tools Dart version 1.25.0-dev.11.0
    • Engine Dart version 2.0.0-dev.9.0

[✓] Android toolchain - develop for Android devices (Android SDK 25.0.3)
    • Android SDK at /Users/hao/Library/Android/sdk
    • Unable to locate Android NDK.

    • Unable to locate compiler in Android NDK.

    • Platform android-25, build-tools 25.0.3
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_112-release-b06)

[✓] iOS toolchain - develop for iOS devices (Xcode 9.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 9.1, Build version 9B55
    • ios-deploy 1.9.2
    • CocoaPods version 1.3.1

[✓] Android Studio (version 2.3)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Java version OpenJDK Runtime Environment (build 1.8.0_112-release-b06)

[✓] Connected devices
    • iPhone 6s • AB7E1A1A-2F6B-469E-BA9F-BFAFA8C9A288 • ios • iOS 11.1 (simulator)

```

### What does this exception mean?

**Comment by eseidelGoogle** [comment]
@goderbauer may have context.

**Comment by goderbauer** [comment]
Can you provide a little more context? What does your code look like that triggered this? Did it only appear after you added something?

**Comment by pulyaevskiy** [comment]
I started getting the same error in my project, however I'm sure I have different context.
I'm still trying to figure out which child is it exactly, but error message is not very helpful.

Can this be extended to include at least runtime type of the child? E.g. "Child with id 25 and type MyObject is not visible..."

Also, probably separate issue: stack traces are not printing out. I thought it was local to my environment, but I see the same issue in the output above:

```
When the exception was thrown, this was the stack:
════════════════════════════════════════════════════════════════════════════════════════════════════
```

**Comment by pulyaevskiy** [comment]
After tinkering with that assertion and changing it to:

```
assert(!child.isInvisible, 'Child with id ${child.id}, ${child.toStringDeep()} is invisible and should not be added to tree.');
```

I got:

```
Child with id 33, SemanticsNode#33[Rect.fromLTRB[0.0, 0.0, 0.0, 0.0]]  is invisible and should not be added to tree.
```

Still need to look more for where it comes from, but a bit better.

Note: I had to set custom `FlutterError.onError` which replaces all `(` and `)` with `[` and `]` before printing to console. Looks like the issue with printing is probably because of a `Zone` specific print function which might be doing something weird with parenthesises.

**Comment by pulyaevskiy** [comment]
Small update:

Not overriding `describeSemanticsConfiguration` with custom configuration in my widget seems to make the error go away.

Here is how mine looked like:

```dart
  @override
  void describeSemanticsConfiguration(SemanticsConfiguration config) {
    super.describeSemanticsConfiguration(config);
    config
      ..value = node.asPlainText
      ..textDirection = textDirection
      ..isFocused = hasFocus
      ..isTextField = true;
  }
```

Typical stack trace:

```
Child with id 52 is invisible and should not be added to tree.
'package:flutter/src/semantics/semantics.dart': Failed assertion: line 370 pos 16: '!child.isInvisible'
#0      _AssertionError._doThrowNew [dart:core-patch/dart:core/errors_patch.dart:37]
#1      _AssertionError._throwNew [dart:core-patch/dart:core/errors_patch.dart:33]
#2      SemanticsNode._replaceChildren [package:flutter/src/semantics/semantics.dart:370:16]
#3      SemanticsNode.updateWith [package:flutter/src/semantics/semantics.dart:628:5]
#4      RenderSemanticsGestureHandler.assembleSemanticsNode [package:flutter/src/rendering/proxy_box.dart:3137:10]
#5      _SwitchableSemanticsFragment.compileChildren [package:flutter/src/rendering/object.dart:3178:13]
#6      _SyncIterator.moveNext [dart:core-patch/dart:core/core_patch.dart:149]
#7      Iterable.single [dart:core/iterable.dart:501]
#8      RenderObject._updateSemantics [package:flutter/src/rendering/object.dart:2263:96]
#9      PipelineOwner.flushSemantics [package:flutter/src/rendering/object.dart:870:16]
#10     BindingBase&SchedulerBinding&GestureBinding&ServicesBinding&RendererBinding.drawFrame [package:flutter/src/rendering/bindin<…>
```

I tried to track what's wrong with the debugger but didn't find any obvious red flags.

The only thing which looked strange was in `RenderSemanticsGestureHandler.assembleSemanticsNode`, lines, 3130-3138:

```dart
    for (SemanticsNode child in children) {
      assert(child.isTagged(useTwoPaneSemantics));
      if (child.isTagged(excludeFromScrolling))
        excluded.add(child); // <--- the bad child ends up here
      else
        included.add(child);
    }
    // node gets updated with excluded list containing "bad" child
    // this tries to replace existing children with excluded list and fails
    node.updateWith(config: null, childrenInInversePaintOrder: excluded);
    _innerNode.updateWith(config: config, childrenInInversePaintOrder: included);
```

Any thoughts, @goderbauer ?

**Comment by goderbauer** [comment]
With the information currently available in this issue I am unable to reproduce this on my end. Can you by any chance post a code example (what does your widget hierarchy look like) that reproduces the problem?

**Comment by lishuhao** [comment]
## My situation
1、I remove old flutter and install flutter in a new directory(different dir name)
2、create a new project and copy old code to this project
3、get dependency and run
4、that exception occur

But when i create a new project and copy part of the old code( which maybe cause this exception) and run ,everything is ok.so i doubt that maybe the env cause this exception

sorry for my poor english

**Comment by lishuhao** [comment]
@goderbauer

**Comment by yathit** [comment]
👍  After the update, I am having similar issue even without overriding `describeSemanticsConfiguration`. It is difficult to reproduce here.

**Comment by yathit** [comment]
```
[✓] Flutter (on Mac OS X 10.12.6 16G1114, locale en-US, channel alpha)
    • Flutter at /Users/kyawtun/flutter
    • Framework revision fd7853faad (5 days ago), 2017-12-05 16:12:55 -0800
    • Engine revision b57fca02b5
    • Tools Dart version 1.25.0-dev.11.0
    • Engine Dart version 2.0.0-edge.d4cfecb1065d322d3670df7e9ec9a0cc2d4b90f0

[✓] Android toolchain - develop for Android devices (Android SDK 26.0.2)
    • Android SDK at /Users/kyawtun/Library/Android/sdk
    • Android NDK location not configured (this optional configuration is useful for native profiling support)
    • Platform android-26, build-tools 26.0.2
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_152-release-915-b08)

[✓] iOS toolchain - develop for iOS devices (Xcode 9.2)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 9.2, Build version 9C40b
    • ios-deploy 1.9.2
    • CocoaPods version 1.3.1

[✓] Android Studio
    • Android Studio at /Applications/Android Studio 3.0 Preview.app/Contents
    • Java version OpenJDK Runtime Environment (build 1.8.0_112-release-b736)

[✓] Android Studio (version 3.0)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Java version OpenJDK Runtime Environment (build 1.8.0_152-release-915-b08)

[✓] IntelliJ IDEA Ultimate Edition (version 2017.3)
    • Flutter plugin version 20.0.3
    • Dart plugin version 173.3727.108

[✓] Connected devices
    • iPhone 7 • 57A902A4-9D2B-48F6-A984-7FDBD170D866 • ios • iOS 11.2 (simulator)
```

The following exception happen when calling ` Navigator.of(context).pushReplacementNamed('/');`.

```
══╡ EXCEPTION CAUGHT BY SCHEDULER LIBRARY ╞═════════════════════════════════════════════════════════
The following assertion was thrown during a scheduler callback:
Child with id 84 is invisible and should not be added to tree.
'package:flutter/src/semantics/semantics.dart': Failed assertion: line 643 pos 16:
'!child.isInvisible'

Either the assertion indicates an error in the framework itself, or we should provide substantially
more information in this error message to help you determine and fix the underlying cause.
In either case, please report this assertion by filing a bug on GitHub:
https://github.com/flutter/flutter/issues/new

When the exception was thrown, this was the stack:
════════════════════════════════════════════════════════════════════════════════════════════════════
Another exception was thrown: Child with id 84 is invisible and should not be added to tree.
)
Another exception was thrown: Child with id 106 is invisible and should not be added to tree.
Another exception was thrown: Child with id 106 is invisible and should not be added to tree.
Another exception was thrown: Child with id 106 is invisible and should not be added to tree.
Another exception was thrown: Multiple widgets used the same GlobalKey.
Another exception was thrown: NoSuchMethodError: The getter 'context' was called on null.
Another exception was thrown: Duplicate GlobalKey detected in widget tree.

```

The exception when re-login after logout. Whenever login `Navigator.of(context).pushReplacementNamed('/');` is called so as to rebuild Home page. Note that `sugarFactor.value` is null after logout and have value after login. Depending on this value Home page is different. 

```

class AppContainer extends StatefulWidget {
  const AppContainer() : super();

  @override
  State<StatefulWidget> createState() => new AppComponentState();
}

class AppComponentState extends State<AppContainer>
    with StoreWatcherMixin<AppContainer> {
  final sugarFactory = new ValueNotifier<Sugar>(null);

  Route _getRoute(RouteSettings settings) {
    if (LOGIN_DEBUG) print('routing to ${settings.name}');
    if (sugarFactory.value == null) {
      if (settings.name == '/') {
        return new MaterialPageRoute<Null>(
          settings: settings,
          builder: (BuildContext context) => new WelcomeScreen(sugarFactory),
        );
      }
    } else {
      RoutePoint rp = new RoutePoint(settings.name);
      if (rp.page == SugarPage.Home) {
        return new MaterialPageRoute<Null>(
          settings: settings,
          builder: (BuildContext context) => new Desktop(sugarFactory.value, desktopStateData),
        );
      } else if (rp.page == SugarPage.Search) {
        return new MaterialPageRoute<Null>(
          settings: settings,
          builder: (BuildContext context) =>
          new SearchPage(sugar, module: rp.mid),
        );
      } else if (rp.page == SugarPage.Module) {
        return new MaterialPageRoute<Null>(
          settings: settings,
          builder: (BuildContext context) => new ModulePage(sugarFactory.value, rp.mid),
        );
      } else if (rp.page == SugarPage.Record) {
        Module m = new Module(sugarFactory.value, rp.mid);
        assert(m != null);
        return new MaterialPageRoute<Null>(
          settings: settings,
          builder: (BuildContext context) => new RecordPage(m, rp.rid),
        );
      }
    }
    return null;
  }

  @override
  Widget build(BuildContext context) {

    return new MaterialApp(
      title: 'App',
      onUnknownRoute: (RouteSettings settings) => new MaterialPageRoute<Null>(
        settings: settings,
        builder: (BuildContext context) => new UnknownPage(settings.name),
      ),
      routes: <String, WidgetBuilder>{
        '/logout': (BuildContext context) => new LogoutPage(sugarFactory),
      },
      onGenerateRoute: _getRoute,
    );
  }
}
```
I used to get longer stack trace previously, but after few release, flutter stack trace are terse and no longer showing stack causing from my app. It is very difficult to debug. Please restore long stack trace as previously. 

![screen shot 2017-12-11 at 1 19 30 am](https://user-images.githubusercontent.com/2137932/33807408-846cb68e-de11-11e7-89cb-5f39cfcf9ccc.png)

**Comment by yathit** [comment]
This happen when Widget have different `key` for same named `MaterialPageRoute`.

**Comment by goderbauer** [comment]
I am sadly still unable to reproduce this with the information given. Can somebody post a complete code example that triggers the `Child with id 84 is invisible and should not be added to tree` assert? Or point to an example app on github that has the problem?

Without that debugging this will be hard.

**Comment by goderbauer** [comment]
I was finally able to reproduce this locally. Will look into fixing it.

**Comment by goderbauer** [comment]
I just submitted the fix for this. Let me know if it works for you after upgrading to the latest master.

**Comment by branflake2267** [comment]
I had the same issue. After upgrade, fixed it. I can confirm this patch worked for me.

**Comment by pulyaevskiy** [comment]
I just updated to latest master and can not reproduce the error (and switching back to alpha makes it re-appear). Looks good so far.

Thanks for fixing this @goderbauer !
